### ISSUE TITLE
[css-shapes][css-images-3] `<radial-size>` syntax seems incorrect

### ISSUE THREAD

**Issue by romainmenke** [w3c/csswg-drafts#10812]
https://drafts.csswg.org/css-images-3/#typedef-radial-size

```
<radial-size> = <radial-extent> | <length [0,∞]> | <length-percentage [0,∞]>{2}
```

This syntax does not allow this:

```css
clip-path: ellipse(closest-side closest-side at 5rem 6rem);
```

Despite that being the implemented feature: https://codepen.io/romainmenke/pen/PordyZx

I think this was a regression in https://github.com/w3c/csswg-drafts/pull/9723

Grouping it seems to resolve the issue:

```
<radial-size> = [ <radial-extent> | <length [0,∞]> | <length-percentage [0,∞]> ]{2}
```

But I am not that familiar with the feature, so my suggestion might be incorrect.

-----

Edit:

The suggested edit would break `circle()`.
I think that that might be the root cause.

That both `circle()` and `ellipse()` are defined with `<radial-size>` while they should be different.

This might work though:

https://drafts.csswg.org/css-images-3/#typedef-radial-size

```diff
- <radial-size> = [ <radial-extent> | <length [0,∞]> | <length-percentage [0,∞]> ]{2}
+ <radial-size> = [ <radial-extent> | <length [0,∞]> | <length-percentage [0,∞]> ]
```

https://drafts.csswg.org/css-shapes/#funcdef-basic-shape-ellipse

```diff
- <ellipse()> = ellipse( <radial-size>? [ at <position> ]? )
+ <ellipse()> = ellipse( [ <radial-size>{2} ]? [ at <position> ]? )
```

**Comment by cdoublev** [comment]
I think you missed the [extension](https://drafts.csswg.org/css-images-4/#radial-size) in CSS Images 4.

But I would also prefer (y)our suggested syntax (#9730).

**Comment by romainmenke** [comment]
Aha, I wasn't aware of the prose in the section. Thank you for finding this 🙇 

Unfortunately we can't parse prose 😛

> Two-component values remain invalid when specifying circle as the [`<radial-shape>`](https://www.w3.org/TR/css-images-3/#typedef-radial-shape), and otherwise indicate the horizontal (first) and vertical (second) radii of the ellipse.

(Working on a CSS declaration linter that uses `@webref/css` as the source of truth.)

**Comment by cdoublev** [comment]
I closed #9730 in favor of this issue, to rephrase my questions about the `<radial-size>` extension in CSS Images 4.

As noted in the comments above, it would be great if the restriction to a single component value for `circle()` and a `circle` gradient shape could be hardcoded into the syntax.

Before this extension:

  - `circle()` could be specified with one `<radial-extent>` (restricted to `*-side`) or `<length-percentage>`
  - a `circle` gradient shape could be specified with one `<radial-extent>` or `<length>`
  - `ellipse()` could be specified with two `<radial-extent>`s (restricted to `*-side`) or `<length-percentage>`s
  - a `ellipse` gradient shape could be specified with one `<radial-extent>` or two `<length-percentage>`s

Now:

  - `circle()` and `circle` gradient shape can be specified with one `<radial-extent>` or `<length-percentage>`
  - `ellipse()` and `ellipse` gradient shape can be specified with one or two `<radial-extent>`s or `<lenth-percentage>`s (possibly mixed, cf. #9729)

Now for `ellipse()` and `ellipse` gradient shape, reading #824, it is not clear whether accepting a single component value was intentional, what the default value (if any) of the second component value and the shortest serialization should be when omitted.

I suspect it should default to the first component value, except when the first is `*-corner`.

Since a single `<length>` (or `<percentage>`, presumably) indicates a `circle` gradient shape, how would you serialize `ellipse 1px` or `ellipse 1px 1px`:

  1. `ellipse 1px`
  2. `1px 1px`

Or maybe it is fine to serialize `1px`? Would it mean that `ellipse(1px)` should serialize as `circle(1px)`?

**Comment by romainmenke** [comment]
FWIW I am now using:

```
<radial-size> = <radial-extent>{1,2} | <length [0,∞]> | <length-percentage [0,∞]>{2}

circle() = circle( [ <radial-extent> | <length-percentage [0,∞]> ]? [ at <position> ]? )
```

**Comment by yisibl** [comment]
Hi @noamr, given that @hyowon-igalia is currently working on this in Blink, could we put this issue on the agenda?
https://chromium-review.googlesource.com/c/chromium/src/+/7754379

**Comment by romainmenke** [comment]
@yisibl or @hyowon-igalia can you summarize the questions there were when implementing this?
And maybe even a proposed way forwards?

It's been a while since I filed this issue and my perspective is more on syntax validation in the context of linters.

**Comment by cdoublev** [comment]
I no longer remember the details of all related issues and resolutions but I kept this grammar as a note:

```
<circle()> = circle( <circle-size>? [ at <position> ]? )
<ellipse()> = ellipse( <ellipse-size>? [ at <position> ]? )
<radial-gradient-syntax> = [ [ circle || <circle-size> ] | [ ellipse || <ellipse-size> ]]?
                           [ at <position> ]?
                           , <color-stop-list>

<circle-size>  = closest-corner | farthest-corner | <radial-radius>
<ellipse-size> = closest-corner | farthest-corner | <radial-radius>{1,2}

<radial-radius> = closest-side | farthest-side | <length-percentage [0,∞]>
```

It could be simplified even further if two `*-corner` can be used to define an ellipse, which I am not sure about (#10833).

`*-corner` also define a radial radius but I could not come up with a better production name than `<radial-radius>` to reuse `*-side` and `<length-percentage>`.

**Comment by hyowon-igalia** [comment]
Hi @romainmenke @cdoublev, the main question was how to handle two independent <radial-extent> keywords in ellipse(), like:

[clip-path: ellipse(farthest-corner closest-corner at 175px 100px);](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/external/wpt/css/css-masking/clip-path/clip-path-ellipse-closest-farthest-corner.html;l=19)

Chromium and WebKit parsers accept <shape-radius>{2}, where each radius resolves independently to the euclidean corner distance - no aspect ratio preservation. The above WPT test also uses this form.

My [CL](https://chromium-review.googlesource.com/c/chromium/src/+/7754379) implemented this based on the current parser behavior to make the WPT test pass, so I hadn't thought deeply about the spec's intent. But looking at the resulting ellipse, to me the current behavior doesn't seem to have a clear geometric meaning - the ellipse doesn't pass through any corner.

If we want an ellipse that passes through the specified corner and fits the box shape naturally, I feel updating the parser and the WPT test to match the spec could be a reasonable direction.

**Comment by yisibl** [comment]
@smfr Since you've implemented this in https://github.com/WebKit/WebKit/pull/31678, we'd like to hear your suggestions.

**Comment by smfr** [comment]
See also https://github.com/w3c/csswg-drafts/issues/14010

**Comment by tabatkins** [comment]
All right, I was asked to provide explictly what the suggested grammar changes would be. I think, first, a bit of history would be helpful, as Images and Shapes have been historically tangled together in this regard and the current state of both specs and impls is confusing if you don't know the order things happened in.

1. Images defined `radial-gradient()` first. Circles could use a single side/corner keyword, or a single length (no %, as it wasn't clear what to resolve it against). Ellipses could use a single side/corner keyword (aspect ratio taken from the reference box), or two length-percentages (resolving %s against width/height).
2. Shapes then defined `circle()` and `ellipse()`, *mostly* copying Images but still doing something unique. `circle()` could take a single *side* keyword (not corner!), or a single length-percentage (%s resolved against `hypot(width, height)`). `ellipse()` took a pair of values, each either a side keyword or a length-percentage. (No way to use the reference box's aspect ratio.)
3. Images 4 then took some lessons from Shapes, relaxing `radial-gradient()` somewhat. Circles could take a length-percentage (%s resolved against `hypot(width, height)`). Ellipse could take two keywords (but deferred the meaning of that to Shapes, meaning two corner keywords was undefined...)
4. Shapes then changed to just lean on Images, switching the grammars to just take `<radial-size>` identical to `radial-gradient()`. This left Images relying on now-missing text from Shapes, and left Shapes missing the additional prose limiting circle() and ellipse() to just the circle/ellipse halves of the grammar.

-------

So, the proposal is that we allow the meaningful union of what these two gramamrs intended to do:

1. Circles can take a single side/corner keyword, or a length-percentage, with %s resolved against `hypot(width, height)`.
2. Ellipses can take 2 side keywords or length-percentages, with %s resolved against width or height. Specifying just one of them is the same as duplicating it. They can alternately specify a single corner keyword, which takes the aspect ratio from the reference box.

Specifically:

```
<circle-size> = closest-side | farthest-side | closest-corner | farthest-corner | <length-percentage>
<ellipse-size> = closest-corner | farthest-corner | [ closest-side | farthest-side | <length-percentage> ]{1, 2}
```

`radial-gradient()` will then pair `<circle-size>` and `<ellipse-size>` with the `circle` and `ellipse` keywords, and define that if the shape keyword is omitted, it's a circle if it matches `<circle-size>` and an ellipse otherwise. `circle()` and `ellipse()` will just use the correct size non-terminal to begin with.

**Comment by hjanuschka** [comment]
Thanks @tabatkins for writing this up, the history makes the current state much easier to understand.

One data point from the implementation side, WebKit has already shipped the two-value corner form for `ellipse()` (WebKit/WebKit#31678 in 2024), i.e. `ellipse(closest-corner farthest-corner)` parses, with each axis resolving independently to the radius it would have if both axes used that keyword (option 2 in #14010, and as far as I understand also what Firefox does). 

Chromium has the same behavior implemented behind a flag, with an approved Intent to Ship that I paused pending this discussion. There is also an existing WPT exercising the two-value corner form (`css/css-masking/clip-path/clip-path-ellipse-closest-farthest-corner.html`).

If I read the proposed grammar correctly, the `{2}` branch of `<ellipse-size>` excludes the corner keywords, which would make that form invalid, so relative to what WebKit ships this would not be purely additive.

Would it be acceptable to keep corners in the two-value form, i.e.

```
<ellipse-size> = <circle-size> | [ closest-side | farthest-side | closest-corner | farthest-corner | <length-percentage> ]{2}
```

with a per-axis corner keyword defined as resolving to the radius it would have if both axes used that keyword?

I realize the geometry is a bit loose (the resulting ellipse may not actually pass through the corner), but it is well-defined, cheap to implement, and matches what is already shipped. 

Everything else in the proposal looks purely additive to me, including the single-value `ellipse()` form and corner keywords for `circle()`.

That said, no strong attachment on my side, if the WG prefers corners to be single-only, I'm happy to update the Chromium implementation and WPTs to match, and to help with spec edits or Gecko/WebKit patches as needed. 

Mostly hoping to get to a resolution so we can unblock shipping.

**Comment by tabatkins** [comment]
I'm *weakly* against including the corner keywords in the two-keyword ellipse form, just because the behavior seems nonsensical to me. You'll only actually hit corners by accident, whereas the side keywords always hit the sides exactly. It makes sense as a "well the spec says so, I guess I can implement something" sort of behavior, but is not something we'd ever design on purpose, I think.

Yeah, looking at the test you've linked, that's definitely just nonsense behavior. The ellipse doesn't touch any of the four corners of the container, because both times it's calculating the radius that would make a *circle* (using the 1:1 aspect ratio of the container) hit the closest or furthest corner, and when you make an ellipse with those axises it fails at both tasks.

So, no, I think we should stick with the proposal above, and only allow side keywords in the two-value ellipse syntax, reserving corner keywords for the single-value syntax where it makes sense.

**Comment by tabatkins** [comment]
(Oh also, not sure what I intended by allowing ellipses to be specified with a single len-per, from the circle-size grammar. In theory we could just duplicate it, like most similar things do, but we specifically *don't* duplicate the side keywords; `closest-side` and `closest-side closest-side` mean different things. So just to keep the single-value terms consistent, I've tweaked the grammar a bit to not reference `<circle-size>` explicitly, and instead write the single-keyword branch explicitly, so it doesn't carry over the single len-per.)

**Comment by hjanuschka** [comment]
sounds good, went ahead and implemented the proposed grammar in Chromium (behind flag) https://chromium-review.googlesource.com/c/chromium/src/+/8427345.

also updated the WPTs, incl. replacing the old two-value corner reftest with single-keyword ones.

one note: kept single <length-percentage> for ellipse() invalid per your last comment, and serialization-wise ellipse(closest-side) collapses to ellipse() while ellipse(farthest-side) expands to the two-value form, happy to adjust once the spec text pins that down.

thank you :) let me know if you want me to address anything, once spec and chromium is done, i am happy to dive into webkit/gecko

**Comment by Loirooriol** [comment]
> `closest-side` and `closest-side closest-side` mean different things

How are they different? Last time I checked, browsers didn't support a single `closest-side` in `ellipse()`. And from the current state of the specs, it's hard to tell.

Also, what specs will be modified? Currently css-images-3 requires a single value (even though no browser allows a single value). So will this just be for css-images-4, or also css-images-3?

**Comment by fantasai** [comment]
Regarding the proposal:

```
<circle-size> = closest-side | farthest-side | closest-corner | farthest-corner | <length-percentage>
<ellipse-size> = closest-side | farthest-side | closest-corner | farthest-corner 
               | [ closest-side | farthest-side | <length-percentage> ]{2}
```

If we're giving single-value `closest-side` the meaning of "like for circles, except with an altered aspect ratio", then I assume we applying the aspect ratio so that the unmeasured axis gets squashed/stretched? If so let's be explicit about that.

The "copy from the aspect ratio" bit is kinda weird imho, but the proposal seems fine under that historical constraint. :)

**Comment by Loirooriol** [comment]
In @hjanuschka's patch I see

```js
test_valid_value("shape-outside", "ellipse(farthest-side)", "ellipse(farthest-side farthest-side)");
```

which seems to imply that if a single value is provided, it gets duplicated, but @tabatkins expected something else.

I'm not really getting the point of allowing a single side value in `ellipse()`, but not allowing a single length.

**Comment by tabatkins** [comment]
> How are they different? Last time I checked, browsers didn't support a single closest-side in ellipse(). And from the current state of the specs, it's hard to tell.

I was referring to `radial-gradient()`, which does allow an ellipse with one keyword, but also I was misremembering what it did. ^_^ Yeah, the side keywords are just repeated; it's only the corner keywords that take the reference box's aspect ratio.

So never mind, I'll go back to allowing a single length-percentage and just duplicating it. It's just the corner keywords that'll have special behavior, *because* they're not allowed in the two-value syntax.

> Also, what specs will be modified?

Certainly Images 4 and Shapes 1, possibly backporting to Images 3.

**Comment by tabatkins** [comment]
Okay, updated. I suppose only remaining question is if we want to allow *explicitly* duplicating the corner keywords (aka `ellipse(closest-corner closest-corner)`) and making all other two-value combos with a corner keyword invalid, or just leave them as single keywords. I lean toward keeping it just a single-value keyword. The grammar is a little silly/awkward to allow it, and it doesn't add any functionality.

**Comment by hjanuschka** [comment]
updated my patch.

one clarification needed on the corner geometry though, you said

> it's only the corner keywords that take the reference box's aspect ratio

but [css-images-3 currently says](https://drafts.csswg.org/css-images-3/#valdef-radial-extent-closest-corner) the corner ellipse gets

> the same aspect-ratio it would have if closest-side were specified

and that's what blink's shipped radial-gradient() does (computes it from the side distances, which makes the corner ellipse exactly the `*-side` ellipse scaled by `sqrt(2)`).

those differ for off-center positions: `300x200` box, center `100px 80px`:

| aspect source | radii |
|---|---|
| side-distance (images-3 prose, blink gradients) | `141.42 / 113.14` |
| reference box | `156.2 / 104.1` |

using the side-distance aspect to stay consistent with gradients, is that the intended definition, or do you really want the box's aspect ratio?
